### PR TITLE
Add clone function for IrGraph.

### PR DESCRIPTION
**Pull request by wzzju**

### CODE REVIEW

File: paddle/fluid/framework/ir/graph.h [hunk 1]
```diff
@@ -71,6 +72,7 @@ namespace ir {
 class Graph {
  public:
   explicit Graph(const ProgramDesc &program);
+  Graph(const Graph &o);
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.h]: can we use a `Clone` method and ban this method.
In the `Clone` comment, say that it only clones graph structure, not the attributes.

Comment (wzzju) [reply paddle/fluid/framework/ir/graph.h]: OK. Thx.

File: python/paddle/fluid/framework.py [hunk 2]
```diff
@@ -1950,6 +1950,19 @@ def __init__(self, graph, for_test=False):
...
+        Create a new and duplicated IrGraph.
+
+        Warns:
+            The method only clones the graph structure, not its attributes.
+
+        Returns:
+            IrGraph: A new and duplicated graph.
+        """
+        g = self.graph.clone()
+        return IrGraph(g, self._for_test)
```
Comment (wanghaoshuang) [inline python/paddle/fluid/framework.py:2016]: `for_test` should be an argument of clone function.

Comment (wzzju) [reply python/paddle/fluid/framework.py:2016]: Hi, I think you should change the `for_test` value when you perform the clone operation. It is different from `Program`.

**Review by panyx0718 — APPROVED** [review]

File: paddle/fluid/framework/ir/graph.cc [hunk 3]
```diff
@@ -152,6 +152,39 @@ void Graph::ResolveHazard(
   }
 }
 
+std::shared_ptr<Graph> Graph::Clone() {
```
Comment (panyx0718) [inline paddle/fluid/framework/ir/graph.cc:155]: no test?

**Review by panyx0718 — CHANGES_REQUESTED** [review]

File: paddle/fluid/framework/ir/graph.cc (hunk 3 above)
Comment (wzzju) [reply paddle/fluid/framework/ir/graph.cc:155]: Have added the executor test for the graph clone API.

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**